lightning/src/ln/peer_handler.rs

```diff
@@ -779,20 +789,29 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 				self.maybe_send_extra_ping(peer);
 			}
 
-			let next_buff = match peer.pending_outbound_buffer.front() {
-				None => return,
-				Some(buff) => buff,
-			};
-
-			let pending = &next_buff[peer.pending_outbound_buffer_first_msg_offset..];
-			let data_sent = descriptor.send_data(pending, peer.should_read());
-			peer.pending_outbound_buffer_first_msg_offset += data_sent;
-			if peer.pending_outbound_buffer_first_msg_offset == next_buff.len() {
-				peer.pending_outbound_buffer_first_msg_offset = 0;
-				peer.pending_outbound_buffer.pop_front();
-			} else {
-				peer.awaiting_write_event = true;
-			}
+			// TODO: break into methods in `peer` or somehow DRY
+			// We prioritize channel messages over gossip broadcasts
+			if let Some(next_buff) = peer.pending_outbound_buffer.front() {
+				let pending = &next_buff[peer.pending_outbound_buffer_first_msg_offset..];
+				let data_sent = descriptor.send_data(pending, peer.should_read());
+				peer.pending_outbound_buffer_first_msg_offset += data_sent;
+				if peer.pending_outbound_buffer_first_msg_offset == next_buff.len() {
+					peer.pending_outbound_buffer_first_msg_offset = 0;
+					peer.pending_outbound_buffer.pop_front();
+				} else {
+					peer.awaiting_write_event = true;
+				}
+			} else if let Some(next_buff) = peer.gossip_broadcast_buffer.front() {
+				let pending = &next_buff[peer.gossip_broadcast_buffer_first_msg_offset..];
+				let data_sent = descriptor.send_data(pending, peer.should_read());
+				peer.gossip_broadcast_buffer_first_msg_offset += data_sent;
+				if peer.gossip_broadcast_buffer_first_msg_offset == next_buff.len() {
+					peer.gossip_broadcast_buffer_first_msg_offset = 0;
+					peer.gossip_broadcast_buffer.pop_front();
+				} else {
+					peer.awaiting_write_event = true;
+				}
+			} else { return }
```

Comment:
If the array is not a problem, I think we could do something with `find_map` (stabilised in 1.30) like (untested):

```suggestion
			if let Some(next_buff) = [peer.pending_outbound_buffer.front(), peer.gossip_broadcast_buffer.front()].iter().find_map(|b| *b) {
				let pending = &next_buff[peer.gossip_broadcast_buffer_first_msg_offset..];
				let data_sent = descriptor.send_data(pending, peer.should_read());
				peer.gossip_broadcast_buffer_first_msg_offset += data_sent;
				if peer.gossip_broadcast_buffer_first_msg_offset == next_buff.len() {
					peer.gossip_broadcast_buffer_first_msg_offset = 0;
					peer.gossip_broadcast_buffer.pop_front();
				} else {
					peer.awaiting_write_event = true;
				}
			} else { return }
```

Reply:
Ah didn't catch that! But latest changes look better anyway 😄